subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/attributes/ArtifactViewAttributesIntegrationTest.groovy

```diff
@@ -0,0 +1,332 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.integtests.resolve.attributes
+
+/**
+ * Tests for [org.gradle.api.artifacts.ArtifactView ArtifactView] that ensure it uses the "live" attributes
+ * of the configuration that created it to select files and artifacts.
+ *
+ * As the necessary attributes are not present on the consumer's compileClasspath configuration until the
+ * beforeLocking callback method is called, the previous incorrect behavior of the ArtifactView was to select without
+ * considering these attributes.  This has been corrected to use a "live" view
+ * of the attributes, to ensure that late changes to the attributes are reflected in the selection.
+ *
+ * Most of these tests use the `java-library` plugin and should resolve the secondary "classes" variant of the producer
+ * and obtain the "main" directory as the only file, instead of the standard jar file variant, which was the legacy
+ * behavior.
+ */
+class ArtifactViewAttributesIntegrationTest extends AbstractArtifactViewAttributesIntegrationTest {
```

Comment:
I feel like this is testing the symptoms too much, which makes it hard to understand what exactly is being tested here.

I don't think we want to test all permutations of creating/resolving artifact views or incoming files.  This was just how we noticed the underlying problem. I think we want a more focused tests for specific behaviors:

1. The results from the "old" and "new" APIs are the same: A test with a consumer and producer. This can use java-library. The consumer has a task that compares the results of incoming files and artifacts and artifact view files and artifacts and verifies that they're the same.  This test would have uncovered the problem because 3/4 of the ways to get files would have been different.  It doesn't really matter which one was first. 
2. The attributes container is "live": A test that adds attributes to a configuration, creates an artifact view and adds more attributes and, finally, the test adds more attributes to the configuration.  When resolving the files for the artifact view, we should see all attributes used. This test explicitly shows that the artifact view can see changes to the configuration's attributes.